bin/service-mango-health/src/processors/persister.rs

```diff
@@ -0,0 +1,392 @@
+                {
+                    log::error!("persistence failed: {}", e);
+                    break;
+                }
+            }
+        });
+
+        let result = PersisterProcessor { job };
+
+        Ok(result)
+    }
+
+    async fn load_previous(client: &Client) -> anyhow::Result<HashMap<Pubkey, PersistedData>> {
+        let rows = client
+            .query(
+                "SELECT Pubkey, Timestamp, healthRatio FROM mango_monitoring.health_current",
+                &[],
+            )
+            .await?;
+
+        let mut result = HashMap::<Pubkey, PersistedData>::new();
+        for row in rows {
+            let key = Pubkey::from_str(row.get(0))?;
+            let ts: chrono::NaiveDateTime = row.get(1);
+            let ts_utc = chrono::DateTime::<Utc>::from_naive_utc_and_offset(ts, Utc);
+            let hr: Option<f64> = row.get(2);
+            result.insert(
+                key,
+                PersistedData {
+                    computed_at: ts_utc,
+                    health_ratio: hr.map(|x| I80F48::wrapping_from_num(x)), // TODO FAS What conversion should we use there ?
+                },
+            );
+        }
+
+        Ok(result)
+    }
+
+    async fn persist(
+        client: &mut Client,
+        previous: &mut HashMap<Pubkey, PersistedData>,
+        event: HealthEvent,
+    ) -> anyhow::Result<()> {
+        let mut updates = HashMap::new();
+
+        for component in &event.components {
+            if !Self::should_insert(&previous, event.computed_at, component.clone()) {
+                continue;
+            }
+
+            updates.insert(
+                component.account,
+                PersistedData {
+                    computed_at: event.computed_at,
+                    health_ratio: component.health_ratio,
+                },
+            );
+        }
+
+        let tx = client.transaction().await?;
+        Self::insert_history(&tx, &updates).await?;
+        Self::update_current(&tx, &updates, &previous).await?; // <- updated & present in previous
+        Self::insert_current(&tx, &updates, &previous).await?; // <- updated & not present in previous
+        Self::delete_old_history(&tx, event.computed_at).await?;
+        tx.commit().await?;
+
+        for (k, v) in updates {
+            previous.insert(k, v);
+        }
+
+        Ok(())
+    }
+
+    async fn insert_history<'tx>(
+        client: &Transaction<'tx>,
+        updates: &HashMap<Pubkey, PersistedData>,
+    ) -> anyhow::Result<()> {
+        for (key, value) in updates {
+            let key = key.to_string();
+            let ts = value.computed_at.naive_utc();
+            let hr = value.health_ratio.map(|x| x.to_num::<f64>());
+            let query = postgres_query::query!("INSERT INTO mango_monitoring.health_history (Pubkey, Timestamp, HealthRatio) VALUES ($key, $ts, $hr)", key, ts, hr);
+            query.execute(client).await.expect("Insertion failed");
```

Comment:
for large number of inserts this is very inefficents, potential steps:
- use tokio_postgres directly, that way we can prepare() the insert statement ahead of time once
- I've seen people use `... VALUES (...row1...), (...row2...), ...` for like 10 rows to manually batch 10 row inserts at a time, but it's really awkward imo
- when looking around, I saw that tokio_postgres supports copy_in(). I think that's supposed to be the fastest way of getting lots of new rows into a table (`https://github.com/sfackler/rust-postgres/issues/681`)

At least the "prepare-statement-only-once" is an optimization I'd like to see.